### PR TITLE
trtllm non causal support

### PR DESCRIPTION
**Pull request by saltyminty**
## 📌 Description

Non-causal (dense-mask) support to trtllm_batch_context_with_kv_cache

NOTE TO REVIEWER: the new "casual" input being inserted in the middle of the public API could cause API regressions to users using positional arguments. I think the current ordering next to window_left makes more sense, but for reviewer to double check if we should instead move it to the end.

## 🔍 Related Issues

https://github.com/flashinfer-ai/flashinfer/issues/2826

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [ ] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [ ] I have installed the hooks with `pre-commit install`.
- [ ] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [ ] Tests have been added or updated as needed.
- [ ] All tests are passing (`unittest`, etc.).

## Summary by CodeRabbit

* **New Features**
  * Public "causal" option for batched prefill so users can select causal vs non-causal attention; runtime execution respects this choice.

* **Bug Fixes / Validation**
  * Added runtime validation: non-causal mode disallowed with sliding-window left offset and requires logits soft-cap == 0.0 (errors raised when violated).

* **Tests**
  * Added and parameterized non-causal attention tests; updated batch-prefill tests and added runtime skips for unsupported combinations.

### CODE REVIEW

File: tests/attention/test_trtllm_gen_attention.py [hunk 1]
```diff
@@ -882,6 +888,57 @@ def test_trtllm_batch_prefill_bs1(
...
 
 
+@pytest.mark.parametrize("kv_layout", ["HND", "NHD"])
+@pytest.mark.parametrize(
+    "batch_size,page_size,num_kv_heads,head_grp_size", [(4, 32, 4, 5)]
+)
+@pytest.mark.parametrize("q_dtype,kv_dtype,o_dtype", [("bf16", "bf16", "bf16")])
+@pytest.mark.parametrize("enable_pdl", [None, False, True])
+@pytest.mark.parametrize("enable_sink", [False, True])
+@pytest.mark.parametrize("max_q_len", [64])
```
Comment (yzh119) [inline tests/attention/test_trtllm_gen_attention.py]: can we add more problem shapes?

File: tests/attention/test_trtllm_gen_attention.py [hunk 2]
```diff
@@ -882,6 +888,57 @@ def test_trtllm_batch_prefill_bs1(
...
+@pytest.mark.parametrize("max_kv_len", [64])
```
Comment (yzh119) [inline tests/attention/test_trtllm_gen_attention.py]: ditto

Comment (saltyminty) [reply tests/attention/test_trtllm_gen_attention.py]: Addressed – now mirrors the existing test cases

File: flashinfer/prefill.py [hunk 3]
```diff
@@ -260,6 +260,7 @@ def _paged_run(
         enable_pdl: bool,
         workspace_size: int,
         window_left: int = -1,
+        is_causal: bool = True,
```
[the remark this answers is not in the document]
Comment (bkryu) [reply flashinfer/prefill.py]: Agree with this comment. We may not want to break API compatibility

File: flashinfer/prefill.py [hunk 4]
```diff
@@ -3741,6 +3744,7 @@ def trtllm_batch_context_with_kv_cache(
     cum_seq_lens_q: torch.Tensor,
     cum_seq_lens_kv: torch.Tensor,
     window_left: int = -1,
+    causal: bool = True,
```
[the remark this answers is not in the document]
Comment (bkryu) [reply flashinfer/prefill.py]: Ditto. Let's keep API compatibility

**Review by bkryu — CHANGES_REQUESTED** [review]
PR looks concise, but breaks API compat. Otherwise looks goot to me. Can you check the comments?

File: flashinfer/prefill.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (saltyminty) [reply flashinfer/prefill.py]: Fixed.

File: flashinfer/prefill.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (saltyminty) [reply flashinfer/prefill.py]: Fixed.

**Comment by bkryu** [comment]
@saltyminty, can you check the internal CI results?

**Comment by saltyminty** [comment]
/bot run

(re-triggering CI after rebase)

**Review by bkryu — APPROVED** [review]
LGTM Pending final internal CI results
